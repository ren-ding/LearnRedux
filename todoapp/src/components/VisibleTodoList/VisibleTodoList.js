import React, {Component} from 'react';
import {StoreContext} from '../../store/store'

class VisibleTodoList extends Component {
    componentDidMount() {
        const {store} = this.context;
        this.unsubscribe = store.subscribe(()=>
            this.forceUpdate()
        );
    }

    componentWillUnmount() {
        this.unsubscribe();
    }

    render() {
        const props = this.props;
        const {store} = this.context;
        const state = store.getState();

        return (
            <TodoList
                todos = {
                    getVisibleTodos(
                        state.todos,
                        state.visibilityFilter
                    )
                }
                onTodoClick = {id =>{
                    store.dispatch({
                        type: 'TOGGLE_TODO',
                        id
                    });
                }}
            />
        );
    }
}


const getVisibleTodos = (
    todos,
    filter
) => {
    switch (filter) {
        case 'SHOW_ALL':
            return todos;
        case 'SHOW_COMPLETED':
            return todos.filter(
                todo => todo.completed
                );
        case 'SHOW_ACTIVE':
            return todos.filter(
                todo => !todo.completed
            );
        default:
            return todos;
    }
};

const Todo = ({
    onClick,
    completed,
    text
}) => (
    <li
        onClick ={onClick}
        style={{textDecoration: completed?
                'line-through':
                'none'}}
    >
        {text}
    </li>
);

const TodoList = ({
    todos,
    onTodoClick
}) => (
    <ul>
        {todos.map(todo => 
            <Todo
                key={todo.id}
                {...todo}
                onClick = {() => onTodoClick(todo.id)}
            />
        )}
    </ul>
);

export default VisibleTodoList;

VisibleTodoList.contextType = StoreContext;