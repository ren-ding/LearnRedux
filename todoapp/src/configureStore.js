import {createStore} from 'redux'
import todoApp from './reducer/index'

// const persistedState = {
//     todos: [{
//         id:'0',
//         text:'Welcome back',
//         completed: false
//     }]
// };

const addLoggingToDispatch = (store) => {
    const rawDispatch = store.dispatch;
    if(!console.group) return rawDispatch;

    return (action) => {
        console.group(action.type);
        console.log('prev state', 'color:gray', store.getState());
        console.log('action','color:blue',action);
        const returnValue = rawDispatch(action);
        console.log('next state','color:green', store.getState());
        console.groupEnd(action.type);
        return returnValue;
    };
};

const configureStore = () => {
    const store = createStore(
        todoApp,
        //persistedState
    );

    if(process.env.NODE_ENV !== 'production') {
        store.dispatch = addLoggingToDispatch(store);
    }

    return store;
}

export default configureStore;