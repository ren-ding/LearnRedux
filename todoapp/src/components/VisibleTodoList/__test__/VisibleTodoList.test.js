import React from 'react';
import renderer from 'react-test-renderer';
import VisibleTodoList from '../VisibleTodoList';
import configureStore from '../../../configureStore';
import {BrowserRouter} from 'react-router-dom';

describe('VisibleTodoList',()=>{ 
    let props;

    beforeEach(()=>{
        props = {
            store: configureStore()
        }
    });

    describe('render',()=>{
        it('should render match with snapshot',()=> {
            const tree = renderer.create(
                <BrowserRouter>
                    <VisibleTodoList {...props}/>
                </BrowserRouter>
            ).toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
});