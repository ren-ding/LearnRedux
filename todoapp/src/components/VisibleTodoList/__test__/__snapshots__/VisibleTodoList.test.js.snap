// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisibleTodoList render should render match with snapshot 1`] = `
<p>
  Loading...
</p>
`;
